modules/table/src/main/java/org/apache/ignite/internal/table/distributed/gc/MvGc.java

```diff
@@ -221,34 +222,62 @@ private void scheduleGcForStorage(TablePartitionId tablePartitionId) {
                 return;
             }
 
-            if (storageHandler.gcInProgressFuture.get() != future) {
+            if (storageHandler.gcInProgressFuture.get() != currentGcFuture) {
                 // Someone in parallel is already collecting garbage, we will try once again after completion of gcInProgressFuture.
                 return;
             }
 
             try {
-                for (int i = 0; i < GC_BATCH_SIZE; i++) {
-                    HybridTimestamp lowWatermark = lowWatermarkReference.get();
+                HybridTimestamp lowWatermark = lowWatermarkReference.get();
 
-                    assert lowWatermark != null : tablePartitionId;
+                assert lowWatermark != null : tablePartitionId;
 
-                    // If storage has been deleted or there is no garbage, then for now we will stop collecting garbage for this storage.
-                    if (!storageHandlerByPartitionId.containsKey(tablePartitionId)
-                            || !storageHandler.storageUpdateHandler.vacuum(lowWatermark)) {
-                        return;
-                    }
-                }
-            } catch (Throwable t) {
-                future.completeExceptionally(t);
+                // If the storage has been deleted, then garbage collection is no longer necessary.
+                if (!storageHandlerByPartitionId.containsKey(tablePartitionId)) {
+                    currentGcFuture.complete(null);
 
-                return;
-            } finally {
-                if (!future.isCompletedExceptionally()) {
-                    future.complete(null);
+                    return;
                 }
-            }
 
-            scheduleGcForStorage(tablePartitionId);
+                StorageUpdateHandler storageUpdateHandler = storageHandler.storageUpdateHandler;
+
+                // We can only start garbage collection when the partition safe time is reached.
+                storageUpdateHandler.getSafeTimeTracker()
+                        .waitFor(lowWatermark)
```

Comment:
How is it possible to have bad safe time inside of the GC task? Shouldn't we only add task to the pool if low watermark is good? Otherwise we end up adding a task that only adds another task, what's the point of it?

Reply:
I don't quite understand what bad safe time means.
Since the safe time is a distributed property of each partition, we have to wait for it, don't we?

The future safe time could complete in a raft thread, so it's safer to move the garbage collection task to the garbage collector's pool.